### DISCUSSION TITLE
Rector conflicts with `nikic/parser` when using phpunit

### DISCUSSION THREAD

**Discussion by nikophil** [rectorphp/rector#8566] [Q&A] [answered]
Hello,

I got a strange behavior where Rector conflicts with `nikic/parser`, but only when running phpunit.

I have a project with the following conf in `composer.json`:
```json
{
    "require": {
        "php": "^8.2",
        "nikic/php-parser": "^5.0"
    },
    "require-dev": {
        "rector/rector": "^1.0",
        "symfony/phpunit-bridge": "6.4.*"
    },
}
```

I can run `composer install` without any problem, as well as accessing to my app via http. Even the project's test suite is OK.
The problem occurs when I want to run rector tests (which inherit from `AbstractRectorTestCase`. 

```
➜ vendor/bin/simple-phpunit --filter AddEmbedResourceSerializationGroupRectorTest
PHPUnit 9.6.17 by Sebastian Bergmann and contributors.

Testing 
PHP Fatal error:  Cannot declare interface PhpParser\Node, because the name is already in use in /app/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/Node.php on line 6
PHP Fatal error:  Uncaught RuntimeException: The autoloader expected class "Composer\InstalledVersions" to be defined in file "/app/vendor/rector/rector/vendor/composer/../composer/InstalledVersions.php". The file was found but the class was not in it, the class name or namespace probably has a typo. in /app/vendor/symfony/error-handler/DebugClassLoader.php:350
Stack trace:
#0 /app/vendor/symfony/error-handler/DebugClassLoader.php(307): Symfony\Component\ErrorHandler\DebugClassLoader->checkClass()
#1 [internal function]: Symfony\Component\ErrorHandler\DebugClassLoader->loadClass()
#2 /app/vendor/symfony/error-handler/DebugClassLoader.php(443): class_exists()
#3 /app/vendor/symfony/error-handler/DebugClassLoader.php(334): Symfony\Component\ErrorHandler\DebugClassLoader->checkAnnotations()
#4 /app/vendor/symfony/error-handler/DebugClassLoader.php(263): Symfony\Component\ErrorHandler\DebugClassLoader->checkClass()
#5 /app/vendor/symfony/phpunit-bridge/DeprecationErrorHandler.php(198): Symfony\Component\ErrorHandler\DebugClassLoader::checkClasses()
#6 [internal function]: Symfony\Bridge\PhpUnit\DeprecationErrorHandler->shutdown()
#7 {main}
  thrown in /app/vendor/symfony/error-handler/DebugClassLoader.php on line 350

Fatal error: Cannot declare interface PhpParser\Node, because the name is already in use in /app/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/Node.php on line 6

Fatal error: Uncaught RuntimeException: The autoloader expected class "Composer\InstalledVersions" to be defined in file "/app/vendor/rector/rector/vendor/composer/../composer/InstalledVersions.php". The file was found but the class was not in it, the class name or namespace probably has a typo. in /app/vendor/symfony/error-handler/DebugClassLoader.php:350
Stack trace:
#0 /app/vendor/symfony/error-handler/DebugClassLoader.php(307): Symfony\Component\ErrorHandler\DebugClassLoader->checkClass()
#1 [internal function]: Symfony\Component\ErrorHandler\DebugClassLoader->loadClass()
#2 /app/vendor/symfony/error-handler/DebugClassLoader.php(443): class_exists()
#3 /app/vendor/symfony/error-handler/DebugClassLoader.php(334): Symfony\Component\ErrorHandler\DebugClassLoader->checkAnnotations()
#4 /app/vendor/symfony/error-handler/DebugClassLoader.php(263): Symfony\Component\ErrorHandler\DebugClassLoader->checkClass()
#5 /app/vendor/symfony/phpunit-bridge/DeprecationErrorHandler.php(198): Symfony\Component\ErrorHandler\DebugClassLoader::checkClasses()
#6 [internal function]: Symfony\Bridge\PhpUnit\DeprecationErrorHandler->shutdown()
#7 {main}
  thrown in /app/vendor/symfony/error-handler/DebugClassLoader.php on line 350
```

I'm also wondering if the problem does not come from Symfony's phpunit bridge, but even when using directly phpunit binary, I got the same error.

NOTE: I saw https://github.com/rectorphp/rector/discussions/6958 but the answers to this discussions seems to not be helpful in my case

any help would be appreciated :)

thanks!

**Comment by samsonasik** [comment]
Could you create simple reproducible repo and step by step to reproduce the error? Thank you.

**Comment by nikophil** [comment]
Hi!

thanks for your answer

here is a reproducer:
https://github.com/nikophil/rector-bug-reproducer

I basically did:
```terminal
$ composer req symfony/flex symfony/framework-bundle
$ composer req test
$ composer req nikic/php-parser
$ composer req --dev rector/rector
$ vendor/bin/rector custom-rule
```

and then:
```terminal
$ vendor/bin/simple-phpunit
PHPUnit 9.6.17 by Sebastian Bergmann and contributors.

Testing 
PHP Fatal error:  Cannot declare interface PhpParser\Parser, because the name is already in use in /home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/Parser.php on line 6

Fatal error: Cannot declare interface PhpParser\Parser, because the name is already in use in /home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/Parser.php on line 6
```

note that `nikic/parser` seems to be the only dep in `vendor/rector/rector/vendor` which seems to be not prefixed. Don't know if this is made on purpose

**Reply by samsonasik** [reply]
Rector is using php-parser 4.x, same with phpstan which rector rely, if you force to use php-parser 5, phpstan will complain:

```
PHP Fatal error:  Class PHPStan\Parser\PhpParserDecorator contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (PhpParser\Parser::getTokens) in phar:///Users/samsonasik/www/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Parser/PhpParserDecorator.php on line 11

Fatal error: Class PHPStan\Parser\PhpParserDecorator contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (PhpParser\Parser::getTokens) in phar:///Users/samsonasik/www/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Parser/PhpParserDecorator.php on line 11
```

I think you can use nikic/php-parser 4.x for now :), once phpstan 2.0 released with php-parser 5.x, rector will follow.

**Reply by nikophil** [reply]
I've updated the project with nikic/php-parser 4.x and the problem remains... (reproducer repository is updated as well).

```shell
$ composer req symfony/flex symfony/framework-bundle
$ composer req test
$ composer req nikic/php-parser:^4.0
$ composer req --dev rector/rector
$ vendor/bin/rector custom-rule
```

and then
```shell
$ vendor/bin/simple-phpunit
PHPUnit 9.6.17 by Sebastian Bergmann and contributors.

Testing 
PHP Fatal error:  Cannot declare interface PhpParser\Parser, because the name is already in use in /home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/Parser.php on line 6

Fatal error: Cannot declare interface PhpParser\Parser, because the name is already in use in /home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/vendor/nikic/php-parser/lib/PhpParser/Parser.php on line 6
```

actually this is the original problem, I did not pay attention the problem was different when I made the reproducer :facepalm:

**Reply by samsonasik** [reply]
That can be due to autoload juggling, running normal `vendor/bin/phpunit` instead of `vendor/bin/simple-phpunit` show error:

```
1) Utils\Rector\Tests\Rector\MyCustomRuleRector\MyCustomRuleRectorTest::test with data set #0 ('/Users/samsonasik/www/rector-...hp.inc')
RuntimeException: The autoloader expected class "Rector\Renaming\Rector\Name\RenameClassRector" to be defined in file "/Users/samsonasik/www/rector-bug-reproducer/vendor/rector/rector/vendor/composer/../../rules/Renaming/Rector/Name/RenameClassRector.php". The file was found but the class was not in it, the class name or namespace probably has a typo.
```

you may need to check `tests/bootstrap.php` autoload ...

**Reply by nikophil** [reply]
ok I got it, but I'm not sure to understand what would be the fix? :pray:

**Reply by samsonasik** [reply]
I have other project, placing `nikic/php-parser` on require-dev, running custom rule seems working ok with `vendor/bin/phpunit`, not sure how `vendor/bin/simple-phpunit` or your `tests/bootstrap.php` and its autoload load it.

**Reply by nikophil** [reply]
I'm sorry, I think I don't understand your answer :sweat: 

even if I completely skip simple-phpunit (meaning, I call directly `vendor/phpunit/phpunit/phpunit` and use `vendor/autoload.php` as bootstrap), the problem still remains:
```shell
➜ vendor/phpunit/phpunit/phpunit --bootstrap vendor/autoload.php 
PHPUnit 9.6.17 by Sebastian Bergmann and contributors.

Testing 
E                                                                   1 / 1 (100%)

Time: 00:00.218, Memory: 40.00 MB

There was 1 error:

1) Utils\Rector\Tests\Rector\MyCustomRuleRector\MyCustomRuleRectorTest::test with data set #0 ('/home/nikophil/works/github.c...hp.inc')
RuntimeException: The autoloader expected class "Rector\Renaming\Rector\Name\RenameClassRector" to be defined in file "/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/vendor/composer/../../rules/Renaming/Rector/Name/RenameClassRector.php". The file was found but the class was not in it, the class name or namespace probably has a typo.

/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/symfony/error-handler/DebugClassLoader.php:354
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/symfony/error-handler/DebugClassLoader.php:311
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/SourceLocator/AutoloadSourceLocator.php:272
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/SourceLocator/FileReadTrapStreamWrapper.php:62
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/SourceLocator/AutoloadSourceLocator.php:266
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/SourceLocator/AutoloadSourceLocator.php:119
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/AggregateSourceLocator.php:26
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/MemoizingSourceLocator.php:33
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/AggregateSourceLocator.php:26
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/SourceLocator/Type/MemoizingSourceLocator.php:33
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/DefaultReflector.php:32
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/Reflector/MemoizingReflector.php:45
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/BetterReflection/BetterReflectionProvider.php:163
phar:///home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/MemoizingReflectionProvider.php:35
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Skipper/SkipVoter/ClassSkipVoter.php:41
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Skipper/Skipper/Skipper.php:59
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/PostRector/Application/PostFileProcessor.php:69
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/PostRector/Application/PostFileProcessor.php:54
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Application/FileProcessor.php:113
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Application/ApplicationFileProcessor.php:184
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Application/ApplicationFileProcessor.php:161
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:202
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:176
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/vendor/rector/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:137
/home/nikophil/works/github.com/nikophil/rector-bug-reproducer/utils/rector/tests/Rector/MyCustomRuleRector/MyCustomRuleRectorTest.php:16
```

**Reply by samsonasik** [reply]
not sure, probably some conflict with symfony deps

**Reply by nikophil** [reply]
Actually, when I comment out [this line](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/ErrorHandler/DebugClassLoader.php#L311), the problem does not occur.

Symfony's error handler seems to be buggy towards this. When I disable `SymfonyTestsListener` everything works!

At the end, in my real project, I'll use another phpunit config file specific for rector.

Thanks a lot for your support! :pray:

**Reply by staabm** [reply]
fwiw, we are diving deeper into this problem in https://github.com/rectorphp/rector/issues/8567#issuecomment-2007626436

**Reply by nikophil** [reply]
super nice, because, in my real life project, I still don't manage to make it work

**Reply by devnix** [reply]
@nikophil had the same problem: uninstall `symfony/phpunit-bridge`

**Comment by ivastly** [comment]
at the time of writing this error is still reproducable (rector 2.1.7, phpstan 2.1.27, nikic/php-parser 5.6.1). Found a dirty fix just to run the tests for my own rector rules - just add `return;` at the first line of `vendor/rector/rector/preload.php`. 

Rector tool itself stops working after this, so I need to always jump between the dirty fix to run the tests and original version to run rector.

**Reply by samsonasik** [reply, marked as answer]
Hi @ivastly , this happen mostly on phpunit 12, you can preload it in composer.json:

```
{
    "autoload-dev": {
        "psr-4": {
            "Your\\Test\\": "tests/"
        },
        "files": [
            "vendor/rector/rector/preload.php"
        ]
    }
}
```

see documentation https://getrector.com/documentation/troubleshooting-preload

**Reply by ivastly** [reply]
works like a charm, thanks a lot! So strange that my LLM has not suggested this fix right away. It proposed to use `rector-src` repo instead.
